File: kakao/2019kakao_prob4.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

vector<int> solution(vector<string> words, vector<string> queries) {
    vector<int> answer;
    map<string, int> map;
    int querySize = queries.size();

    for(auto iter = queries.begin(); iter != queries.end(); iter++) {
        string query = *iter;
        string keyWord = "";
        int queryLen = query.length();
        bool isTail = (query[0] != '?');

        if(map.find(query) != map.end()) {
            answer.push_back(map[query]);
            continue;
        }

        for (int i = 0; i < queryLen; i++)
        {
            if (query[i] != '?')
            {
                keyWord += query[i];
            }
        }

        int count = 0;
        for(auto witer = words.begin(); witer != words.end(); witer++) {
            string str = *witer;
            if(str.length() == queryLen) {
                string subStr;

                if(isTail) {
                    subStr = str.substr(0, keyWord.length());
                } else {
                    subStr = str.substr(queryLen - keyWord.length());
                }
                
                if(subStr == keyWord) {
                    count++;
                }
            }
        }
        map.insert({query, count});
        answer.push_back(count);
    }

    return answer;
}

int main()
{
    vector<int> res = solution({"frodo", "front", "frost", "frozen", "frame", "kakao"}, {"fro??", "????o", "fr???", "fro???", "pro?", "???????", "fro??"});
    for(int t : res){
        cout << t << " ";
    }
    return 0;
}

File: kakao/2019kakao_prob4.js
function solution(words, queries) {
    var answer = [];

    queries.forEach(query => {
        let keyWord = "";
        const queryLen = query.length;
        const isTail = (query[0] != '?');

        for(var index = 0; index < queryLen; index++) {
            if(query[index] != '?') {
                keyWord += query[index];
            }
        }

        let count = 0;
        words.sort();
        words.forEach(word => {
            if(word.length === queryLen) {
                let subStr;
                
                if(isTail) {
                    subStr = word.substr(0, keyWord.length);
                } else {
                    subStr = word.substr(queryLen - keyWord.length);
                }

                if(subStr > keyWord){
                    return;
                }
                
                if(subStr === keyWord) {
                    count++;
                }
            }
        });

        answer.push(count);
    });

    return answer;
}

const res = solution(["frodo", "front", "frost", "frozen", "frame", "kakao"], ["fro??", "????o", "fr???", "fro???", "pro?", "???nt"]);
console.log(res);